src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,341 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "uiScreens.h"
+
+enum nativeScriptType {
+	NATIVE_SCRIPT_PUBKEY = 0,
+	NATIVE_SCRIPT_ALL = 1,
+	NATIVE_SCRIPT_ANY = 2,
+	NATIVE_SCRIPT_N_OF_K = 3,
+	NATIVE_SCRIPT_INVALID_BEFORE = 4,
+	NATIVE_SCRIPT_INVALID_HEREAFTER = 5,
+};
+
+// distinguish between the types of PUBKEY
+// DEVICE_OWNED is specified by a path
+// THIRD_PARTY is specified by a pubkey hash
+enum {
+	PUBKEY_DEVICE_OWNED = 0,
+	PUBKEY_THIRD_PARTY = 1,
+};
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+bool expectsRequest()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->remainingScripts[ctx->level] > 0;
+}
+
+bool shouldFinish()
+{
+	// we finish if there are no more scripts to be processed
+	return ctx->level == 0 && ctx->remainingScripts[0] == 0;
+}
+
+static inline void scriptFinished()
+{
+	--ctx->remainingScripts[ctx->level];
+	TRACE("native script finished, level = %u, remainingScripts = %u", ctx->level, ctx->remainingScripts[ctx->level]);
+}
+
+static inline void complexScriptFinished()
+{
+	--ctx->level;
+	scriptFinished();
+}
+
+// UI
+
+static void deriveScriptHash_display_ui_callback()
+{
+	if (shouldFinish()) {
+		TRACE("whole native script received");
+		ui_idle();
+		// TODO: hashing, this is just a mock response
```

Comment:
Hashing is needed after each APDU, not just once in the end. Generally, we first do computations (e.g. hashing) and only afterwards call ui operations.